Fe enriched overlapping domains

niveshd opened the pull request:
ColorEnriched::Helper class provided in "FE_Enriched.h" to create hp::FECollection, a collection of finite element objects used by hp::DoFHandler, in a domain with multiple, possibly overlapping, sub-domains with individual enrichment functions. Note that the overlapping regions may have multiple enrichment functions associated with it. This is implemented using a general constructor of FE_Enriched object which allows different enrichment functions and graph coloring functions.

davydden commented:
/run-tests

davydden commented on doc/news/changes/minor/20180420NiveshD:
```diff
@@ -0,0 +1,9 @@
+New: ColorEnriched::Helper class provided to create hp::FECollection, a 
```
IMO this deserves to be a `major` category

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
 
+
+
+namespace ColorEnriched
```
you need to document the namespace for doxygen, I think.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+      (const typename Triangulation<dim>::cell_iterator &) >;
+
+  /**
+   * ColorEnriched::Helper class provided to create hp::FECollection, a
```
class creates a collection of finite elements (hp::FECollection) to be used with hp::DoFHandler...

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * using a general constructor of FE_Enriched object which allows different
+   * enrichment functions.
+   *
+   * <code> Helper<dim> </code> is first constructed by providing a base
```
~~first~~

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   *
+   * <code> Helper<dim> </code> is first constructed by providing a base
+   * FiniteElement @p fe_base, the enrichment FiniteElement @p fe_enriched,
+   * a vector of predicate functions and enrichment functions
```
...predicate functions that define subdomains as well as the corresponding enrichment functions.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * <code> Helper<dim> </code> is first constructed by providing a base
+   * FiniteElement @p fe_base, the enrichment FiniteElement @p fe_enriched,
+   * a vector of predicate functions and enrichment functions
+   * corresponding to subdomains. The class can then, by using a member
```
> The class can then

something is wrong here, please reword. 
The logic should be that this class populates FECollection that a user can retrieve and use in hp::DoFHandler.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * \htmlonly <style>div.image img[src="3source_fe_indices.png"]{width:25%;}</style> \endhtmlonly
+   * @image html 3source_fe_indices.png "Active FE indices"
+   *
+   * The name of FE element at an index i of hp::FE_Collection @p fe_collection
```
FE ~~element~~

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * The name of FE element at an index i of hp::FE_Collection @p fe_collection
+   * can be obtained by <code>fe_collection[index].get_name()</code> and is
+   * show in the table below. Note all the FE enriched elements are of same size.
+   * For a cell with fe index 1 and another cell with fe index 2, the enrichement
```
FE

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+  struct Helper
+  {
+    /**
+     * A simple constructor for Helper class.
```
~~A simple constructor for Helper class.~~ Constructor.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+     */
+    Helper(const FE_Q<dim> &fe_base,
+           const FE_Q<dim> &fe_enriched,
+           const std::vector<predicate_function<dim>> &vec_predicates,
```
vec_predicates -> predicates

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+    Helper(const FE_Q<dim> &fe_base,
+           const FE_Q<dim> &fe_enriched,
+           const std::vector<predicate_function<dim>> &vec_predicates,
+           const std::vector<std::shared_ptr<Function<dim>>> &vec_enrichments);
```
vec_enrichments -> enrichments

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+
+    /**
+     * Prepares @p dof_handler by modifying material id of cells and
+     * constructs data members cellwise_color_predicate_map and fe_sets which
```
do not expose internal implementation. in `public: ` we shall just explain what this does and make sure users understand when this is to be called. How is it implemented internally and which private members are being initialized here is not important.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,160 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+    hp::FECollection<dim> get_fe_collection() const;
+
+  private:
+    hp::FECollection<dim> fe_collection;
```
you need to document all `private:` members.

davydden commented:
@niveshd you miss headers in the tests:
```
Building CXX object CMakeFiles/fe_enriched_color_07.debug.dir/fe_enriched_color_07.cc.o
/home/bob/source/tests/fe/fe_enriched_color_07.cc:261:3: error: â€˜CSplineâ€™ in namespace â€˜dealii::Functionsâ€™ does not name a type
   Functions::CSpline<1> cspline;
```
and this
```
2659/3561 Test #2660: fe/fe_enriched_color_02.debug ........................................***Failed   14.48 sec
Test fe/fe_enriched_color_02.debug: RUN
===============================   OUTPUT BEGIN  ===============================
Generating fe_enriched_color_02.debug/interrupt_guard.cc
Scanning dependencies of target fe_enriched_color_02.debug
Building CXX object CMakeFiles/fe_enriched_color_02.debug.dir/fe_enriched_color_02.cc.o
Building CXX object CMakeFiles/fe_enriched_color_02.debug.dir/fe_enriched_color_02.debug/interrupt_guard.cc.o
Linking CXX executable fe_enriched_color_02.debug/fe_enriched_color_02.debug
Built target fe_enriched_color_02.debug
Scanning dependencies of target fe_enriched_color_02.debug.diff
Generating fe_enriched_color_02.debug/output
fe/fe_enriched_color_02.debug: BUILD successful.
fe/fe_enriched_color_02.debug: RUN failed. ------ Return code 134
fe/fe_enriched_color_02.debug: RUN failed. ------ Result: /home/bob/build-gcc/tests/fe/fe_enriched_color_02.debug/failing_output
fe/fe_enriched_color_02.debug: RUN failed. ------ Partial output:

--------------------------------------------------------
An error occurred in line <493> of file </home/bob/source/tests/fe/../tests.h> in function
    MPILogInitAll::MPILogInitAll(bool)
The violated condition was: 
    false
Additional information: 
    This exception -- which is used in many places in the library -- usually indicates that some condition which the author of the code thought must be satisfied at a certain point in an algorithm, is not fulfilled. An example would be that the first part of an algorithm sorts elements of an array in ascending order, and a second part of the algorithm later encounters an element that is not larger than the previous one.

There is usually not very much you can do if you encounter such an exception since it indicates an error in deal.II, not in your own program. Try to come up with the smallest possible program that still demonstrates the error and contact the deal.II mailing lists with it to obtain help.
--------------------------------------------------------

Aborted (core dumped)
make[3]: *** [fe_enriched_color_02.debug/output] Error 1
make[2]: *** [CMakeFiles/fe_enriched_color_02.debug.diff.dir/all] Error 2
make[1]: *** [CMakeFiles/fe_enriched_color_02.debug.diff.dir/rule] Error 2
make: *** [fe_enriched_color_02.debug.diff] Error 2


fe/fe_enriched_color_02.debug: ******    RUN failed    *******
```
for `fe_enriched_color_05`, `fe_enriched_color_04` and the rest. If they really need MPI you should add to the name of `.output` files `.with_mpi` so that they are disabled for no-MPI config.

niveshd commented:
For graph coloring, i am using Zoltan library which requires mpi initialization even if i am running in serial. So how about i just add .with_mpi=true to all tests that need this graph coloring?

davydden commented:
> So how about i just add .with_mpi=true to all tests that need this graph coloring?

ok, but please mention this in the documentation of the class.

niveshd commented on tests/fe/fe_enriched_color_07.cc:
```diff
@@ -0,0 +1,2038 @@
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/parsed_function.h>
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/function_cspline.h>
```
Regarding error - tests/fe/fe_enriched_color_07.cc

I already had the header for cspline. The error doesn't show up in my runs!

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,231 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * Different cells are associated with FE indices as shown in the following
+   * image.
+   * \htmlonly <style>div.image img[src="3source_fe_indices.png"]{width:25%;}</style> \endhtmlonly
+   * @image html 3source_fe_indices.png "Active FE indices"
```
please continue paragraph by explaining connectivity graph  graph and its partitioning which ultimately leads to how we populate FECollection

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,231 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * @image html 3source_fe_indices.png "Active FE indices"
+   *
+   * The name of finite element at an index (i) of @p fe_collection
+   * (hp::FE_Collection) can be obtained by
```
FECollection

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,231 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+     */
+    std::vector<std::shared_ptr <Function<dim>> > enrichments;
+
+    using cell_iterator_function
```
please document this.

niveshd commented:
@davydden Thanks for the feedback. i modified the functions and class to take spacedim parameter

davydden commented:
@dealii/dealii can someone look into this FE_Enriched related PR? The author is my student so I should not really approve it by myself.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * different for the two sub-domains, it is handled by using
+   * dominating FE_Nothing object (a dummy enriched FiniteElement).
+   * This is similar to how interface between enriched and non-enriched
+   * domains is handled. It is ensured that both sub-domains have same number of
```
have the same number

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * enrichment functions (dummy or not). Another complication is when
+   * they have different (kind of) enrichment functions
+   * (Say the corresponding FE_Enriched objects are fe_enr1, fe_enr2).
+   * This is handled by pairing them with a dominating FE_Nothing i.e
```
i.e.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * domains is handled. It is ensured that both sub-domains have same number of
+   * enrichment functions (dummy or not). Another complication is when
+   * they have different (kind of) enrichment functions
+   * (Say the corresponding FE_Enriched objects are fe_enr1, fe_enr2).
```
do not capitalize "Say"

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * This is handled by pairing them with a dominating FE_Nothing i.e
+   * use FE_Enriched<dim>[fe_base, fe_enr1, fe_nothing] and
+   * FE_Enriched<dim>[fe_base, fe_nothing, fe_enr2]. Note that
+   * size of vector of enriched FiniteElements is equal to 2, the
```
Note that the size of the vector

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * use FE_Enriched<dim>[fe_base, fe_enr1, fe_nothing] and
+   * FE_Enriched<dim>[fe_base, fe_nothing, fe_enr2]. Note that
+   * size of vector of enriched FiniteElements is equal to 2, the
+   * same as number of enrichment functions. Using the above two
```
as the number of

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * same as number of enrichment functions. Using the above two
+   * strategies, we construct the appropriate FE_Enriched
+   * using the general constructor. Note that this is
+   * done on mesh without hanging nodes.
```
on a mesh

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * be the same as number of enrichment functions, possibly
+   * using a number of dummy FE_Nothing objects. This is because
+   * we are not using the information of how the domains are connected
+   * and any sub-domain may share interface with another (Not
```
do not capitalize "Not"

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * and any sub-domain may share interface with another (Not
+   * considering overlaps for now!).
+   * However, in general, a given sub-domain
+   * shares interface only with a few sub-domains. This warrants
```
shares the interface

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * considering overlaps for now!).
+   * However, in general, a given sub-domain
+   * shares interface only with a few sub-domains. This warrants
+   * the use graph coloring algorithm to reduce
```
the use of a graph coloring algorithm

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * However, in general, a given sub-domain
+   * shares interface only with a few sub-domains. This warrants
+   * the use graph coloring algorithm to reduce
+   * the size of vector of enriched FiniteElements
```
the size of the vector

By the way, I would avoid using *FiniteElements* here and elsewhere and instead write *FiniteElement objects* or *finite elements*. The first alternative will insert the link to the `FiniteElement` class, whereas your spelling will just end up looking a bit weird.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * the use graph coloring algorithm to reduce
+   * the size of vector of enriched FiniteElements
+   * (Used in the FE_Enriched constructor). By giving the sub-domains
+   * that share no interface a single color, a single enrichment
```
a single color -> the same color

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * that share no interface a single color, a single enrichment
+   * function that returns different enrichment functions for each
+   * sub-domain can be constructed. Then the vector size of enriched
+   * FiniteElements is equal to number of different colors
```
equal to the number of

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * enrichment functions), a vector of predicate
+   * functions that define sub-domains as well as the corresponding enrichment
+   * functions. The class can then, by using a member
+   * function, set(hp::DofHandler &), modify active FE indices of mesh cells.
```
modify the active FE indices

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * Consider a domain with three sub-domains defined by predicate functions.
+   * Different cells are associated with FE indices as shown in the following
+   * image. The three equal-sized square-shaped sub-domains can be seen.
+   * The 1'st and 2'nd sub-domain intersect with cell labelled with FE
```
labelled -> labeled

Furthermore, I would simply write 1st, 2nd, and 3rd.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * The 1'st and 2'nd sub-domain intersect with cell labelled with FE
+   * index 3. The 3'rd sub-domain is on the left-down corner labelled with FE
+   * index 1. As can be seen, connections exist between sub-domains 1 and 2,
+   * 2 and 3 but 1'st and 3'rd domains are not connected.
```
but the 1st and 3rd

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * @image html 3source_fe_indices.png "Active FE indices"
+   *
+   * As discussed before, the colors of predicates are alloted using
+   * a graph coloring algorithm. Each predicate is a node in graph and if
```
a node in a graph

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * two sub-domains share interface, the corresponding predicates
+   * should be given different colors.
+   * (The predicate colors are different
+   * from what is shown in image. The colors in image are as per FE indices).
```
shown in the image

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * (The predicate colors are different
+   * from what is shown in image. The colors in image are as per FE indices).
+   * Predicates 1 and 3 can be given the same color since they
+   * are not connected but color given to 2'nd
```
given to the 2nd

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * (hp::FECollection) can be obtained by
+   * <code>fe_collection[index].get_name()</code> and is
+   * show in the table below. Note that all the FE_Enriched elements
+   * are of same size and FE_Nothing<2>() is used as discussed before.
```
are of the same size

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   *
+   * The internal data members used by this class need to be available when the
+   * problem is solved. This can be ensured by declaring the object static,
+   * which then remains in the memory until the end of program. An alternative
```
the run of the program ends.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+           const std::vector<std::shared_ptr<Function<spacedim>>> &enrichments);
+
+    /**
+     * Prepares @p a hp::DoFHandler object. The active FE indices of
```
Remove stray `@p` here. `hp::DoFHandler` will get picked up and rendered as a link by doxygen automatically, by the way.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+     *
+     * @return hp::FECollection<dim>, a collection of finite elements
+     */
+    hp::FECollection<dim,spacedim> get_fe_collection() const;
```
Is this intentionally that you return a copy of the FE here and not just a reference (or a `unique_ptr` of a clone)?

niveshd replied:
This was intentional. I could have made it return a const reference but i didn't want to leak the implementation. I thought it would be more error safe as well this way.

bangerth replied:
Copying a whole FE Collection object is a *really really* expensive operation. Let's try not to do that.

masterleinad replied:
What about a `std::shared_ptr`? That sounds like the best option if your class should still have (shared) ownership.

bangerth replied:
Shared pointers only make sense if you expect that (i) it's unknown which side will keep the pointer for longer, or (ii) either side can modify the data.

I have not read through your code, so I don't know whether that's the case. But if all you want is to expose a member so others can see it -- like, for example, `DoFHandler::get_fe()` --, then returning a `const` reference is the way to go.

kronbichler commented on include/deal.II/fe/fe_enriched.templates.h:
```diff
@@ -0,0 +1,206 @@
+  {
+    /**
+     * Returns true if there is a connection between subdomains in the mesh
+     * associated with @p hp::DoFHandler i.e if the subdomains share at least a
```
i.e.,

kronbichler commented on include/deal.II/fe/fe_enriched.templates.h:
```diff
@@ -0,0 +1,206 @@
+     * Returns true if there is a connection between subdomains in the mesh
+     * associated with @p hp::DoFHandler i.e if the subdomains share at least a
+     * vertex. The two subdomains are defined by predicates provided by
+     * @p predicate_1 and @p predicate_2. A predicate is a a function (or
```
A predicate is a function (remove one stray 'a')

kronbichler commented on include/deal.II/fe/fe_enriched.templates.h:
```diff
@@ -0,0 +1,206 @@
+     * object of a type with an operator()) which takes in a cell iterator and
+     * gives a boolean. It is said to be active in a cell if it returns true.
+     *
+     * An example of a custom predicate is one that checks distance from a fixed
```
checks the distance from

kronbichler commented on include/deal.II/fe/fe_enriched.templates.h:
```diff
@@ -0,0 +1,206 @@
+     *
+     * @param[in] hp::DoFHandler object
+     * @param[in] predicate_1 A function (or object of a type with an operator())
+     * defining the subdomain 1. The function takes in an cell and returns a
```
takes in a cell

niveshd replied:
Thanks a lot for all remarks!

kronbichler commented on include/deal.II/fe/fe_enriched.templates.h:
```diff
@@ -0,0 +1,206 @@
+
+namespace ColorEnriched
+{
+  namespace internal
```
Is there any reason why you put only declarations in the `fe_enriched.templates.h` file? We use `.templates.h` files for definitions, i.e., implementations of templated functions that a user might want to add its own instantiations for. I don't see that here. Is the code simply restricted to the implementation in `fe_enriched.cc`? In that case, simply move these internal declarations there. Or, if there needs to be a global visibility, I would put them into `fe_enriched.h`

niveshd replied:
Thanks for looking through. I needed the declarations in header files because some of the tests verify the functions in "fe_enriched.templates.h" individually. I didn't want them in the header file "fe_enriched.h" since they are implementation details.

niveshd replied:
@kronbichler I made the changes you suggested. Please review them. 

I haven't yet moved the functions since i didn't get a reply regarding this. But let me know if you prefer them moved. I will make the change.

masterleinad commented:
Please rebase this PR onto the current master soon to make sure this PR is conforming to the new indenting style we are using. If you need help, don't hesitate to ask.

niveshd commented:
@masterleinad I rebased the PR. To correct the style, I downloaded the clang format and ran the "contrib/utilities/indent ." I hope that's the right procedure. This modified the file "source/physics/elasticity/kinematics.inst.in" as well which i just ignored because its outside my edits.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * same for the sub-domains, FE_Enriched object of one sub-domain
+   * is constructed such that each enriched FE is paired (figuratively) with a
+   * FE_Nothing in the FE_Enriched object of the other sub-domain.
+   * For example, let enriched FEs fe_enr1, fe_enr2 correspond
```
let the FEs \`fe_enr1\`, \`fe_enr2\` , which will be used with enrichment functions, correspond...

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * For example, let enriched FEs fe_enr1, fe_enr2 correspond
+   * to the two sub-domains. Then the FE_Enriched objects of the two
+   * sub-domains are
+   * FE_Enriched<dim,spacedim>[fe_base, fe_enr1, fe_nothing] and
```
> build using \'[fe_base, fe_enr1, fe_nothing]\'

(current text is a bit misleading as it suggests there is such constructor in FE_Enriched)

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * to the two sub-domains. Then the FE_Enriched objects of the two
+   * sub-domains are
+   * FE_Enriched<dim,spacedim>[fe_base, fe_enr1, fe_nothing] and
+   * FE_Enriched<dim,spacedim>[fe_base, fe_nothing, fe_enr2] respectively.
```
> \`[fe_base, fe_nothing, fe_enr2]\`

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * the use of a graph coloring algorithm to reduce
+   * the size of the vector of enriched FEs
+   * (used in the FE_Enriched constructor). By giving the sub-domains
+   * that share no interface the same color, a single enrichment
```
a single \`std::function\`

(and then remove `function` on the next line)

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * two sub-domains share an interface, the corresponding predicates
+   * should be given different colors.
+   * (The predicate colors are different from what is shown
+   * in the image. The colors in the image are as per FE indices).
```
colors in the image represent different FE indices (?)

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * which is deallocated only when the program terminates. An alternative
+   * would be to use it as a data member of the containing class. Since vector
+   * of predicates and enrichment functions may not be available while
+   * constructing the Helper, a shared_ptr to Helper object can be used and
```
\'std::shared_ptr\'

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+     * @return shared pointer to hp::FECollection, a collection of
+     * finite elements needed by @p dof_handler.
+     */
+    std::shared_ptr<hp::FECollection<dim, spacedim>>
```
i thought you wanted to return a `const hp::FECollection<dim, spacedim>&` ?

niveshd replied:
i wanted to but FECollection only has 
operator= (FECollection< dim, spacedim > &&)=default.

So i just used shared pointer.

bangerth replied:
Then you probably forgot to add the ampersand at the end. You may not be able to return it as a value, but you should always be able to return it as a `const` reference.

davydden replied:
ok

niveshd replied:
i don't have the commits, but I think I used const reference &. The function return was not an issue but when I wrote 
(FECollection) = (function return) 
It says it couldn't find the function operator= for const reference & argument!

I will check again and get back to you.

masterleinad replied:
Then, you are invoking the copy constructor somewhere.

bangerth replied:
Yes -- you need to use the result to initialize a reference variable, not later to copy by value. So this will work:
```
  const FECollection &fe_collection = object.get_fe_collection(...);
```
because it initializes a reference variable. But this will fails:
```
  FECollection fe_collection;
  fe_collection = object.get_fe_collection(...);
```
because it copies a value, which you can't do here because `FECollection` does not have a copy operator.

niveshd replied:
Now i see what you meant. I will try changing it.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+     * <code>@p predicates[i]</code> returns true for a cell if it
+     * belongs to a sub-domain with index (i).
+     */
+    std::vector<predicate_function<dim, spacedim>> predicates;
```
i think that can be `const std::vector<predicate_function<dim, spacedim>>&`?

niveshd replied:
Thanks. I think that would be better. I will try.

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -1046,6 +1048,378 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+  template <int dim, int spacedim>
+  std::shared_ptr<hp::FECollection<dim, spacedim>>
+  Helper<dim, spacedim>::get_fe_collection(
+    hp::DoFHandler<dim, spacedim> &dof_handler)
```
By the way, I think the function name is poorly chosen. A function called `get_...` should simply return something. But this function takes a non-`const` DoFHandler object, computes all sorts of things, and at the end returns an FE collection. I think naming this function better for its purpose would be useful :-)

davydden replied:
probably `build_fe_collection()` would better.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1046,6 +1048,443 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+
+
+      /*
+       * Treat interface between enriched cells specially
```
full stop

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1046,6 +1048,443 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+
+      /*
+       * Treat interface between enriched cells specially
+       * Since the function
```
reword it something like:

Until https://github.com/dealii/dealii/issues/1496 is resolved, each time we building constraints at the interface
between two different FE_Enriched, we look for the least dominating FE via hp::FECollection< dim, spacedim >::find_least_face_dominating_fe(). If we don't take further actions, we may find a dominating FE that is too
restrictive, i.e. enriched FE consisting of only....

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1046,6 +1048,443 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+       * enrichment functions 1 and 2 be represented by [1 1 0], with the last
+       * entry as zero since the 3rd enrichment function is not relevant for
+       * the cell. If the interface has enriched FE [1 0 1] and [0 1 1]
+       * on adjacent cells, the an enriched FE [0 0 1] should exist and is
```
~~an~~

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1046,6 +1048,443 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+       */
+
+      // loop through faces
+      for (auto cell = dof_handler.begin_active(); cell != dof_handler.end();
```
let's keep this approach (cell loop) for now and first make sure this solves your problems. 

Ideally, i think you don't have to loop over faces, instead you can go through edges in your graph (which is already built based on geometric arguments) and then check if for a pair of nodes for a given edge you need to add new FE to FECollection.

niveshd replied:
The graph only has information about the connections between sub-domains. The intersections may or may not exist and the main problem is between intersections and the pure sub-domains. so i am looping through all the faces.

niveshd commented:
@kronbichler I moved the functions from "fe_enriched.templates.h" to "fe_enriched.h" and removed the template file. I made the other changes you asked. Please let me know what you think.

niveshd commented:
@davydden  sorry that was a temporary commit (I needed it for some test runs in a different computer!). Now it should be fine.

The modified algo doesn't run into the error regarding the dominating FE element. The constraints generated look good now.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+       * If we don't take further actions, we may find a dominating FE that
+       * is too restrictive, i.e. enriched FE consisting of only FE_Nothing.
+       * New elements needs to be added to FECollection object to help
+       * find the correct enriched FE element underlying the spaces in the
```
~element~

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+       */
+
+      // loop through faces
+      for (auto cell = dof_handler.begin_active(); cell != dof_handler.end();
```
use range loop

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+        for (unsigned int face = 0; face < GeometryInfo<dim>::faces_per_cell;
+             ++face)
+          {
+            unsigned int fe_index = cell->active_fe_index();
```
const, move outside of `face` loop.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+            // find active indices of neighboring cells
+            if (!cell->at_boundary(face))
+              {
+                unsigned int nbr_fe_index =
```
const

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+             ++face)
+          {
+            unsigned int fe_index = cell->active_fe_index();
+            auto         fe_set   = fe_sets[fe_index];
```
use `std::map::at()` , move outside of `face` loop, result is const.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+                unsigned int nbr_fe_index =
+                  cell->neighbor(face)->active_fe_index();
+                // find corresponding fe set
+                auto nbr_fe_set = fe_sets[nbr_fe_index];
```
use `std::map::at()`. Also the result should be `const` (you won't modify it for this neighbour)

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+            if (!cell->at_boundary(face))
+              {
+                unsigned int nbr_fe_index =
+                  cell->neighbor(face)->active_fe_index();
```
before doing his, you should visit each face only once, to that end you can, for example, use material id which you set above, i.e.
```
if (cell->set_material_id < nbr_cell->material_id)
   ... do something
```

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+       * entry as zero since the 3rd enrichment function is not relevant for
+       * the cell. If the interface has enriched FE [1 0 1] and [0 1 1]
+       * on adjacent cells, an enriched FE [0 0 1] should exist and is
+       * found as the dominating finite element for the two cells by
```
...as the **least** dominating...

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+
+                // find intersection of the fe_sets
+                std::set<unsigned int> intersection_set;
+                for (auto s : fe_set)
```
```
std::set_intersection(fe_set.begin(),fe_set.end(), nbr_fe_set.begin(), nbr_fe_set.end(),
                  std::inserter(intersection_set, intersection_set.begin()));
```
see https://stackoverflow.com/a/13448094/888478

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+                bool found = false;
+                if (!intersection_set.empty())
+                  {
+                    for (unsigned int j = 0; j < fe_sets.size(); ++j)
```
use
```
const auto it = std::find(fe_sets.begin(), fe_sets.end(), intersection_set);
if (it != fe_sets.end())
  {
  }
```
https://en.cppreference.com/w/cpp/algorithm/find

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+          bool found = false;
+          if (!color_list.empty())
+            {
+              for (unsigned int j = 0; j < fe_sets.size(); ++j)
```
you can also use `std::find` here. You can get `j` via `std::distance`.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,7 +696,502 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+     * 5 (color = 2). cellwise_color_predicate_map has this information
+     * provided we know the material id.
+     *
+     * Now a call to color_enrichment[1](cell) should in turn call
```
The constructed `color_enrichments` is such that
`color_enrichments[color](cell)` will return return a pointer to function with id=4, i.e. enrichments[4].
In other words, using the previously collected information in this function we translate a vector of user provided enrichment functions into a vector of functions suitable for FE_Enriched class.

davydden replied:
@niveshd i think you missed this one.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+          color_enrichments[i] =
+            [&, i](const typename Triangulation<dim, spacedim>::cell_iterator
+                     &cell) {
+              unsigned int id = cell->material_id();
```
const

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+              unsigned int id = cell->material_id();
+
+              /*
+               * i'th color_enrichment function corresponds to (i+1)'th color.
```
why this i+1 ? Is it related to Zoltan?

niveshd replied:
Yes. The Zoltan coloring starts from 1.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+      // loop through color sets and create FE_enriched element for each
+      // of them provided before calling this function, we have color
+      // enrichment function associated with each color.
+      for (unsigned int color_set_id = 0; color_set_id != fe_sets.size();
```
~!=~ <

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+            {
+              // Given a color id ( = *it), corresponding color enrichment
+              // function is at index id-1 because color_enrichments is
+              // is indexed from zero.
```
double `is`

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+            num_colors, &fe_nothing);
+          EnrichmentFunctions_2DVector functions(num_colors, {dummy_function});
+
+          for (auto it = fe_sets[color_set_id].begin();
```
range loop.
```
for (const auto unshifted_ind : fe_sets[color_set_id])
  {
  }
```

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+               ++it)
+            {
+              // Given a color id ( = *it), corresponding color enrichment
+              // function is at index id-1 because color_enrichments is
```
the same question with that +1/-1 shift, what's the origin?

niveshd replied:
If i want color function corresponding to an index i (say = 3), the corresponding color function is (i-1) i.e 2.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+        return nullptr;
+      };
+
+      using EnrichmentFunctions_2DVector =
```
there is only one place where you use it, so i would not add that `using` shortcut.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+      // define dummy function which is associated with FE_Nothing
+      using cell_function = std::function<const Function<spacedim> *(
+        const typename Triangulation<dim, spacedim>::cell_iterator &)>;
+      cell_function dummy_function;
```
one-liner:

```
const std::function<const Function<spacedim> *(
  const typename Triangulation<dim, spacedim>::cell_iterator &)> 
dummy_function = 
[=]...
```

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+
+              // color_set_id'th color function is (color_set_id-1)
+              // element of color wise enrichments
+              functions[ind].assign(1, color_enrichments[ind]);
```
i think 
```
funcitons[ind][0] = color_enrichments[ind];
```
is more readable as you already have the right size of single element with default being `dummy_function`

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+    , fe_nothing(1, true)
+    , predicates(predicates)
+    , enrichments(enrichments)
+  {}
```
assert that predicates and enrichments are of the same size.

davydden replied:
assert that `fe_base.n_components() == fe_enriched.n_components()`

davydden replied:
you can also assert that size of predicates is positive.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+    const std::vector<std::shared_ptr<Function<spacedim>>> &enrichments)
+    : fe_base(fe_base)
+    , fe_enriched(fe_enriched)
+    , fe_nothing(1, true)
```
~1~ fe_base.n_components()

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+  {
+    // color the predicates based on connections between corresponding
+    // subdomains
+    if (predicates.size() != 0)
```
remove `if` given that you add `Assert` that the size is positive.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+                                                          fe_nothing,
+                                                          fe_collection);
+
+    return fe_collection;
```
that's a concise function 😄

davydden commented:
while doing minor clean-ups, you can of course check that no tests are broken. To be on the safe side, do not squash the commit with others (current state that you use yourself).

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+                  cell->set_active_fe_index(fe_sets.size() - 1);
+                }
+            }
+          ++map_index; // map_index should be unique to cells
```
maybe add explanation why is that. If I recall correctly, it's something like:

At the end of the day, we will have only a handful of combinations of colors stored in `fe_sets`. However two distant cell that have the same combination of colors may be using different enrichment functions for such colors (which is the raison d'etre for this algorithm). This information is stored in `cellwise_color_predicate_map`.

davydden commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -691,7 +696,502 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   *                                             enrichments);
+   * const hp::FECollection<dim>&
+   * fe_collection(FE_helper.build_fe_collection(dof_handler));
+   * @endcode
```
At the end of doxygen part u can add ‘ @author bla-bla’

niveshd commented on source/fe/fe_enriched.cc:
```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+          // children cells. So, the cellwise_color_predicate_map stays
+          // relevant.
+          cell->set_material_id(map_index);
+          std::set<unsigned int> color_list;
```
I don't rely on the order. Actually if its not ordered, i will have issues when i true to find (in the current algo.) if the fe set already existed.

davydden replied:
> Actually if its not ordered, i will have issues

then add somewhere a note (where you check if a given combination exists) that we rely on the set of colors being ordered, i.e. we will have {1,3} but not {3,1}.

jppelteret commented:
@kronbichler Are you satisfied with the changes made in this PR? Can we merge this?

davydden commented:
> @kronbichler Are you satisfied with the changes made in this PR? Can we merge this?

@kronbichler ping.
